### ISSUE TITLE
export2sql dont include 'groupchat' and 'users' table

### ISSUE THREAD

**Issue by oko235** [processone/ejabberd#3680]
## Environment

- ejabberd version: 21.07
- Erlang version: Erlang (SMP,ASYNC_THREADS) (BEAM) emulator version 10.2.4
- OS: Linux (Debian 10)
- Database : mnesia to postgresql
- Installed from: source

## Configuration (extract only about sql settings) ejabberd.yml

```yaml
auth_method: [sql]
default_db: sql
new_sql_schema: true
#"db_type: sql" in modules :
mod_announce
mod_caps
mod_last
mod_muc
mod_mam
mod_offline
mod_privacy
mod_pubsub
mod_roster
mod_shared_roster
mod_vcard
```

## Errors from error.log/crash.log

Not any errors

## Bug description

When I use "ejabberdctl export2sql domain.tld /path/file.sql", sql file dont include "groupchat" in archive table (only kind='chat') and "users" table.
Using "ejabberdctl dump /path/dump.xml" seems include everything, "users" and "groupchat" this time.

Unable to connect with empty users table (auth_method: [sql] seems good). To connect and test muc mam configuration, I've parse "dump.xml" to extract users accounts and put into postgres database. I join muc with gajim and write messages, messages well put in archive table.
So, according to my configuration set to use sql instead of mnesia database, I dont understand why export2sql dont export that elements.

Tested too with "ejabberdctl live" then "ejd2sql:export(<<"domain.tld">>, sql)." Good transfers into postgres database but same result and without any error.

**Comment by badlop** [comment]
@Neustradamus Why do you add that heart icon? Did you test the problem yourself? Did you suffer the problem yourself? Do you provide any valuable effort yourself?

If you do not provide any value to this problem, please remove that icon, as it is confusing.

I will not investigate this problem until you remove that useless icon.

**Comment by badlop** [comment]
> When I use "ejabberdctl export2sql domain.tld /path/file.sql", sql file dont include "groupchat" in archive table (only kind='chat') 

Right! It's a bug in ejd2sql.erl I've fixed it now in https://github.com/processone/ejabberd/commit/cfc393a12e3507c9ecb541b8f784c14ff3a9786e

> and "users" table.

I tried to reproduce that problem, but the feature works correctly for me. Maybe you didn't set sql as auth_method?

I'll explain how I test it, maybe this gives you some clue about where the problem is.

I install ejabberd 21.07, and define a new vhost like this:

```erlang
hosts:
  - localhost
  - pgsql.localhost
```

Start ejabberd, register a new account in that vhost called pg1. It is stored in Mnesia.

Stop ejabberd

Configure the vhost to use PGSQL, something like
```erlang
    sql_port: 5432
    sql_type: pgsql
    sql_server: "pgsql"
    sql_database: ...
    sql_username: ...
    sql_password: ...
    sql_pool_size: 1
    auth_method: sql
    default_db: sql
    default_ram_db: sql
```

Start ejabberd and:
```
❯ ejabberdctl dump /tmp/aa.txt
❯ ejabberdctl export2sql pgsql.localhost /tmp/bb.sql
```

aa.txt shows:
```erlang
{passwd,{<<"pg1">>,<<"pgsql.localhost">>},<<"asd">>}.
```
bb.sql shows:
```sql
-- 
-- Mnesia table: passwd
--
delete from users where username='pg1' and 0=0;
INSERT INTO users(username, password) VALUES ('pg1', 'asd');
```

**Comment by oko235** [comment]
Thank you very much for bug confirmation and fix about groupchat mam !

About users, I tested like you with an empty database. You're right, it works correctly for me too, in txt dump and sql export with "auth_method: sql | internal" and "auth_password_format=plan | scram" always works.

But when I use my data, not any users account in sql export.
I tested with my data and register new account via "ejabberdctl register testuser2 domain.tld testuser2" or via gajim. It works in sql export but _only_ for that new account not old ones.

insert from sql export : 
```
INSERT INTO users(username, server_host, password, serverkey, salt, iterationcount)
VALUES ('testuser1', 'domain.tld', 'Imc698f+2bg/7NLlcg7N9B4qdn4=', 'RECe+FC8zmnERhtoogp9DMJkzjY=', 'gjmtMy7usn0ME19OdI1MSQ==', 4096);
```
and from txt dump : 
```
{passwd,{<<"testuser1">>,<<"domain.tld">>},
        {scram,<<"W549tgwgq4Z34XjJk1mAHMJvAUY=">>,
               <<"HdhLMCFQdLhVSgqwox1WD3sZc80=">>,
               <<"P71vc6z/zDMbphk6QZoXxQ==">>,sha,4096}}.
```
When I compare entry in dump file, I can see "sha" that is not present in old account
```
{passwd,{<<"olduser">>,<<"domain.tld">>},
        {scram,<<"m3Zd6s3Om1JXesQr/Wg5a9fo284=">>,
               <<"mNjGpH8HQALeupEU7GZ30Euzm0g=">>,
               <<"2/+faH8NeVy6XL86mBJdVA==">>,4096}}.
```
It's probably a clue to explain the differences between old and new account and maybe why old account dont be exported in sql.

**Comment by badlop** [comment]
Aha!

xmpp's scram includes the hash mechanism (by default 'sha') since commit https://github.com/processone/xmpp/commit/651050f9

ejabberd 20.12 uses that xmpp, and most of ejabberd code was updated to support both old and new scram records (https://github.com/processone/ejabberd/commit/e5cad9be6 and other commits)... but ejd2sql.erl didn't, so it supports only the new one.

I've fixed that.

In your case, if you can't apply the patch, one dirty workaround:
1. Dump the passwd mnesia table to a text file with https://docs.ejabberd.im/developer/ejabberd-api/admin-api/#dump-table
> ejabberdctl dump_table /tmp/old.txt passwd
2. Edit that text file, adding "sha," to all the records, so those records use the new scram.
3. Load the modified text file (it will overwrite existing entries, no need to delete them before)
> ejabberdctl load /tmp/new.txt
4. Now you can finally use export2sql
> ejabberdctl export2sql pgsql.localhost /tmp/db.sql

**Comment by oko235** [comment]
Thanks again for new fix !

get my sql script now and ready to import in postgresql database.

But new problem or my mistake maybe on a subtlety : 
from `psql -h localhost -p 5432 -U xmpp -d domain -f domain.sql`
with `standard_conforming_strings` set to `off` in `postgresql.conf` (no default value but seen in ejabberd documentation).
Postgresql return me some errors `WARNING:  nonstandard use of escape in a string literal at character xx` about `\n` and `\"`
Example at : 
```
INSERT INTO vcard(username, server_host, vcard)
VALUES ('usertest1', 'domain.tld', '<vCard xmlns=''vcard-temp''>\n<FN>UserTest1</FN>\n</vCard>');
```
or
```
INSERT INTO muc_room(name, host, server_host, opts)
VALUES ('roomname', 'conf.domain.tld', 'domaint.tld', '[…]',{password,<<\"secretpassword\">>},[…],{affiliations,[{{<<\"usertest1\">>,<<\"domain.tld\">>,<<>>},{owner,<<>>}}]},{subject,[]},{subject_author,<<>>},{subscribers,[]}]');
```

If I switch `standard_conforming_strings` to `on`, perfect loading in postgres but when I start ejabberd get : 
```
[error] <0.363.0>@ejabberd_sql:decode_term:284 Corruped Erlang term in SQL database:
** Scanner error: at line 1: unterminated string starting with "log\">>},{anonymo"
```
and probably problems on line break in archive and co with client usage.

Did I miss something or do I have to substitute `-e 's/\\"/"/g' -e 's/\\n/\n/g'` in sql script ?

**Comment by badlop** [comment]
This problem goes outside my knowledge, I can just offer what I got:

```sql
delete from vcard where username='admin' and 0=0;
INSERT INTO vcard(username, vcard) VALUES ('admin', '<vCard xmlns=''vcard-temp''><N><GIVEN>mi primer...

delete from muc_room where name='sala1' and host='conf.localhost';
INSERT INTO muc_room(name, host, opts) VALUES ('sala1', 'conf.localhost', '[{lang,<<\"es\">>},...
```

It's curious that vcard gets two single-quotes: `''` , but muc_room gets one double-quote: `"`

I get those warnings too
```
pgsql      | 2021-10-01 17:09:55.218 UTC [735] WARNING:  nonstandard use of escape in a string literal at character 75
```

Anyway, import in PgSQL 13.2 from docker goes correctly, ejabberd starts, user can login and gets his vcard

**Comment by oko235** [comment]
Thanks again for your answer.

Last question before closing my issue with success.

It's more by curiosity than technical but, on my production server : `archive_msg.DAT` size 394M, after restoring backup on my test server `archive_msg.DAT` size 110M. How to explain the difference ?

**Comment by badlop** [comment]
You can check a mnesia table size (in rows and memory) in ejabberd web admin -> Nodes -> your node -> Database, and check if they match in old and new server. I bet they will match, and the overhead in the old DAT file is caused by some garbage not yet collected, or checkpoints, or other runtime stuff.

I found no specific clue in https://erlang.org/doc/apps/mnesia/Mnesia_chap7.html#startup-files,-log-file,-and-data-files
